fedimintd/src/bin/main.rs

```diff
@@ -72,7 +75,43 @@ async fn main() -> anyhow::Result<()> {
         registry.init();
     }
 
-    let mut task_group = TaskGroup::new();
+    let mut root_task_group = TaskGroup::new();
+    root_task_group.install_kill_handler();
+
+    let task_group = root_task_group.clone();
+    let (main_shutdown_tx, main_shutdown_rx) = tokio::sync::oneshot::channel::<()>();
+
+    root_task_group
+        .spawn_local("main", move |_task_handle| async move {
+            match run(opts, task_group).await {
+                Ok(()) => {
+                    info!("Shutting down");
+                }
+                Err(e) => {
+                    error!(?e, "Main task returned error, shutting down");
+                }
+            }
+
+            main_shutdown_tx
+                .send(())
+                .expect("Could not send shutdown signal");
+        })
+        .await;
+
+    main_shutdown_rx.await.expect("Main task panicked");
+
+    root_task_group.shutdown().await;
```

Comment:
As mentioned above  - this should be `join_all`, and we don't want limit the time out `shutdown()` (if finishes right away), but the time of `join_all` **after** `is_shutting_down` flag was set.